stetho/src/main/java/com/facebook/stetho/inspector/elements/ShadowDOM.java

```diff
@@ -0,0 +1,397 @@
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.inspector.elements;
+
+import com.facebook.stetho.common.Accumulator;
+import com.facebook.stetho.common.ListUtil;
+import com.facebook.stetho.common.Util;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+public final class ShadowDOM implements DOMView {
+  private final Object mRootElement;
+  private final IdentityHashMap<Object, ElementInfo> mElementToInfoMap = new IdentityHashMap<>();
+  private boolean mIsUpdating;
+
+  public ShadowDOM(Object rootElement) {
+    mRootElement = Util.throwIfNull(rootElement);
+  }
+
+  @Override
+  public Object getRootElement() {
+    return mRootElement;
+  }
+
+  @Override
+  public ElementInfo getElementInfo(Object element) {
+    return mElementToInfoMap.get(element);
+  }
+
+  public UpdateBuilder beginUpdate() {
+    if (mIsUpdating) {
+      throw new IllegalStateException();
+    }
+
+    mIsUpdating = true;
+
+    return new UpdateBuilderImpl();
+  }
+
+  public static interface UpdateBuilder {
+    public void setElementChildren(Object element, List<Object> children);
+
+    public Update build();
+  }
+
+  public static interface Update extends DOMView {
+    public boolean isEmpty();
+
+    public boolean isElementChanged(Object element);
+
+    public void getChangedElements(Accumulator<Object> accumulator);
+
+    public void getGarbageElements(Accumulator<Object> accumulator);
+
+    public void abandon();
+
+    public void commit();
+  }
+
+  private final class UpdateBuilderImpl implements UpdateBuilder {
+    /**
+     * We use a {@link LinkedHashMap} to preserve ordering between
+     * {@link UpdateBuilder#setElementChildren(Object, List)} and
+     * {@link Update#getChangedElements(Accumulator)}. This isn't needed for correctness but it
+     * significantly improves performance.
+     *
+     * Transmitting DOM updates to Chrome works best if we can do it in BFS order because it allows
+     * us to skip processing (and, more importantly, transmission) of an element that was already
+     * transmitted in a previous DOM.childNodeInserted event (i.o.w. we can skip transmission of E2
+     * if it was already bundled up in E1's event, where E2 is any element in E1's sub-tree).
+     * DOM.childNodeInserted transmits the node being inserted by-value, so it takes time and space
+     * proportional to the size of that node's sub-tree. This means the difference between O(n^2)
+     * and O(n) time for transmitting updates to Chrome.
+     *
+     * Since {@link com.facebook.stetho.inspector.elements.android.AndroidDOMProvider} already
```

Comment:
Tell IntelliJ to collapse this.  I haven't yet modified Stetho's proejct settings to do it for you :)